crates/storage/src/cache.rs

```diff
@@ -1,252 +1,150 @@
+                Err(e)
             }
-        };
-
-        // Fill in the lock state and send down the complete tx.
-        *slot_lock = SlotState::Ready(res.clone());
-        if complete_tx.send(res.clone()).is_err() {
-            warn!("failed to notify waiting cache readers");
+            Err(e) => Err(DbError::Other(e.to_string())),
         }
-
-        Ok(res)
     }
 
-    /// Returns a clone of an entry from the cache or invokes some function to load it from
-    /// the underlying database.
     pub(crate) fn get_or_fetch_blocking(
         &self,
-        k: &K,
+        key: &K,
         fetch_fn: impl Fn() -> DbResult<V>,
     ) -> DbResult<V> {
-        // The flow control here is kinda weird, I don't like it.  The key here is that we want to
-        // ensure the lock on the whole cache is as short-lived as possible while we check to see if
-        // the entry we're looking for is there.  If it's not, then we want to insert a reservation
-        // that we hold a lock to and then release the cache-level lock.
-        let (mut slot_lock, complete_tx) = {
-            let mut cache = self.cache.lock();
-            if let Some(entry_lock) = cache.get(k).cloned() {
-                drop(cache);
-                let entry = entry_lock.blocking_read();
-                return entry.get_blocking();
-            }
-
-            // Create a new cache slot and insert and lock it.
-            let (complete_tx, complete_rx) = broadcast::channel(1);
-            let slot = Arc::new(RwLock::new(SlotState::Pending(complete_rx)));
-            cache.push(k.clone(), slot.clone());
-            let lock = slot
-                .try_write_owned()
-                .expect("cache: lock fresh cache entry");
-
-            (lock, complete_tx)
-        };
-
-        // Load the entry and insert it into the slot we've already reserved.
-        let res = match fetch_fn() {
-            Ok(v) => v,
-            Err(e) => {
-                warn!(?e, "failed to make database fetch");
-                *slot_lock = SlotState::Error;
-                self.purge(k);
-                return Err(e);
-            }
-        };
-
-        // Fill in the lock state and send down the complete tx.
-        *slot_lock = SlotState::Ready(res.clone());
-        if complete_tx.send(res.clone()).is_err() {
-            // This happens if there was no waiters, which is normal, leaving it
-            // here if we need to debug it.
-            //warn!("failed to notify waiting cache readers");
+        if let Some(v) = self.get(key) {
+            return Ok(v);
         }
-
-        Ok(res)
+        block_in_place(move || {
+            runtime::Handle::current().block_on(async {
+                self.get_or_fetch(key, || {
+                    let (tx, rx) = oneshot::channel();
+                    let _ = tx.send(fetch_fn());
+                    rx
+                })
+                .await
+            })
+        })
```

Comment:
This assumes that there is a runtime context on the current thread.  This means that it does not work on main or on threads started with `thread::spawn`, which is all of them (unless you do a `Handle::enter` call).

The locking primitives have blocking functions.  This is so that you can use them without being in a runtime context.  That's what the original version did.

Reply:
No, that's not the correct behavior since that involves a setting up [a *ton* of bookkeeping (kernel calls, allocations)](https://docs.rs/tokio/latest/src/tokio/runtime/scheduler/current_thread/mod.rs.html) to set up the runtime and make the call, which also immediately blocks the runtime, only to tear it down immediately after.

The correct behavior is to use the functions designed for this use-case.  This also goes a long way to ensure that we avoid *any* additional context switches in un-uncontended contexts.